### PR TITLE
feat(permissions): add ACL attribute based-access control checking for video uploads

### PR DESCRIPTION
**Pull request by joelhooks**
requires this on backend -> https://github.com/eggheadio/egghead-rails/pull/4848

uses -> https://www.npmjs.com/package/accesscontrol

rolls

### CODE REVIEW

File: src/pages/api/aws/sign-s3.ts [hunk 1]
```diff
@@ -20,28 +24,38 @@ const s3 = new AWS.S3(options)
 
 const signedUrl = async (req: NextApiRequest, res: NextApiResponse) => {
   if (req.method === 'GET') {
-    const {objectName, contentType} = req.query
+    const user = await loadCurrentUser(req.cookies[ACCESS_TOKEN_KEY])
 
-    const filename = `${uuidv4()}/${objectName}`
+    const roles = intersection(user.roles, allRoles)
```
Comment (joelhooks) [inline src/pages/api/aws/sign-s3.ts:29]: when we load the roles from the server there are some exrtras so we do an "intersection" with the ones we are using here in this app to remove those extras and not throw an error

File: src/server/accesscontrol.ts [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+  'pro',
+  'instructor',
+  'editor',
+  'admin',
+  'publisher',
+  'editor',
+  'instructor',
+  'reviewer',
+  'advocate',
+]
```
Comment (joelhooks) [inline src/server/accesscontrol.ts:16]: all the roles we care about

File: src/server/accesscontrol.ts [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+
+accessControl.grant('user')
+accessControl.grant('pro').extend('user')
+accessControl.grant(['editor', 'reviewer']).extend('pro')
+accessControl.grant(['advocate']).extend('reviewer')
+accessControl.grant(['instructor']).extend('pro')
+accessControl.grant('publisher').extend(['editor', 'instructor'])
+accessControl.grant('admin').extend(['editor', 'instructor', 'publisher'])
```
Comment (joelhooks) [inline src/server/accesscontrol.ts:24]: kind of goofy, but they have other better ways, just wanted to get it in for now

File: src/server/accesscontrol.ts [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+
+accessControl.grant('admin').createOwn('video')
```
Comment (joelhooks) [inline src/server/accesscontrol.ts:26]: we only have one thing for now!

**Comment by joelhooks** [comment]
this library looks a lot more polished. 

https://github.com/stalniy/casl

based on the same principles as cancan

File: src/server/ability.ts [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+  if (user.roles?.includes('admin')) {
+    can('manage', 'all') // read-write access to everything
+  }
+
+  if (user.roles?.includes('admin', 'editor', 'publisher')) {
+    can('upload', 'Video')
+  }
+
+  return build()
+}
```
Comment (joelhooks) [inline src/server/ability.ts:15]: very similar to cancan and extremely robust in terms of how we can define permissions

https://casl.js.org/v5/en/guide/define-rules

File: src/server/ability.ts [hunk 6]
```diff
@@ -0,0 +1,19 @@
+import {AbilityBuilder, Ability} from '@casl/ability'
+
+type Actions = 'manage' | 'upload'
+type Subjects = 'Video' | 'all'
+type AppAbility = Ability<[Actions, Subjects]>
```
Comment (joelhooks) [inline src/server/ability.ts:5]: very cool TS support

https://casl.js.org/v5/en/advanced/typescript

File: src/server/accesscontrol.ts (hunk 2 above)
Comment (joelhooks) [reply src/server/accesscontrol.ts]: 🤮

File: src/server/accesscontrol.ts (hunk 3 above)
Comment (joelhooks) [reply src/server/accesscontrol.ts]: 🚮

File: src/server/accesscontrol.ts (hunk 4 above)
Comment (joelhooks) [reply src/server/accesscontrol.ts]: 👎

File: src/server/ability.ts (hunk 5 above)
Comment (joelhooks) [reply src/server/ability.ts]: I refactored this to be nice and tidy. Very happy with how it turned out.

File: src/components/search/instructors/kyle-shevlin/index.tsx [hunk 7]
```diff
@@ -18,8 +18,6 @@ const SearchKyleShevlin = ({instructor}: {instructor: any}) => {
 
   const [primaryCourse, ...restCourses] = courses.resources
 
-  console.log(restCourses)
```
Comment (joelhooks) [inline src/components/search/instructors/kyle-shevlin/index.tsx:21]: there were a bunch of these logs around so I removed them.

File: src/server/ability.ts [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+
+export async function getAbilityFromToken(token: string) {
+  const user = await loadCurrentUser(token)
+  return defineAbilityFor(user)
+}
+
+function defineAbilityFor(user: any) {
+  const {can, build} = new AbilityBuilder<AppAbility>(Ability)
```
Comment (jbranchaud) [inline src/server/ability.ts:14]:
```suggestion
  // AbilityBuilder https://casl.js.org/v5/en/guide/define-rules#ability-builder-class
  const {can, build} = new AbilityBuilder<AppAbility>(Ability)
```

Comment (jbranchaud) [reply src/server/ability.ts:14]: The `AbilityBuilder` API is a little indirect, so I thought it'd be handy to link to the docs right above it.

**Review by jbranchaud — APPROVED** [review]

**Pull request merged**